### ISSUE TITLE
OSError: [Errno 84] Invalid or incomplete multibyte or wide character: b'\xff'

### ISSUE THREAD

**Issue by dotlambda** [ronf/asyncssh#547]
```
============================= test session starts ==============================
platform linux -- Python 3.10.10, pytest-7.2.1, pluggy-1.0.0
rootdir: /build/asyncssh-2.13.1, configfile: tox.ini, testpaths: tests
collected 1502 items / 32 deselected / 1470 selected

tests/test_agent.py ...................                                  [  1%]
tests/test_asn1.py .                                                     [  1%]
tests/test_auth.py ........                                              [  1%]
tests/test_auth_keys.py .........                                        [  2%]
tests/test_channel.py .................................................. [  5%]
.....................................................                    [  9%]
tests/test_compression.py .                                              [  9%]
tests/test_config.py ................................................... [ 13%]
.......                                                                  [ 13%]
tests/test_connection.py ................................ss............. [ 16%]
........................................................................ [ 21%]
....s...................................................                 [ 25%]
tests/test_connection_auth.py .......................................... [ 28%]
........................................................................ [ 33%]
........................................................................ [ 38%]
.............................                                            [ 40%]
tests/test_editor.py ....................                                [ 41%]
tests/test_encryption.py ..                                              [ 41%]
tests/test_forward.py .................................................. [ 44%]
.........................................................                [ 48%]
tests/test_kex.py ..........s                                            [ 49%]
tests/test_known_hosts.py ................                               [ 50%]
tests/test_logging.py .........                                          [ 51%]
tests/test_mac.py .s                                                     [ 51%]
tests/test_packet.py ..                                                  [ 51%]
tests/test_pkcs11.py ............                                        [ 52%]
tests/test_process.py .................................................. [ 55%]
.............ssssss.............                                         [ 57%]
tests/test_public_key.py ................                                [ 59%]
tests/test_saslprep.py .......                                           [ 59%]
tests/test_sftp.py ................F.................................... [ 63%]
........................................................................ [ 68%]
........................................................................ [ 72%]
...........................F............................................ [ 77%]
........................................................................ [ 82%]
........................................................................ [ 87%]
...............................................................          [ 91%]
tests/test_sk.py ......................                                  [ 93%]
tests/test_stream.py ......................                              [ 94%]
tests/test_subprocess.py ............                                    [ 95%]
tests/test_x11.py ......................................                 [ 98%]
tests/test_x509.py .........................                             [100%]

=================================== FAILURES ===================================
_______________________ _TestSFTP.test_copy_invalid_name _______________________

self = <tests.test_sftp._TestSFTP testMethod=test_copy_invalid_name>
sftp = <asyncssh.sftp.SFTPClient object at 0x7ffff4a56440>

    @sftp_test
    async def test_copy_invalid_name(self, sftp):
        """Test copying a file with an invalid name over SFTP"""

        for method in ('get', 'put', 'copy', 'mget', 'mput', 'mcopy'):
            with self.subTest(method=method):
                with self.assertRaises((FileNotFoundError, SFTPNoSuchFile,
                                        SFTPFailure, UnicodeDecodeError)):
>                   await getattr(sftp, method)(b'\xff')

/build/asyncssh-2.13.1/tests/test_sftp.py:873:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/build/asyncssh-2.13.1/asyncssh/sftp.py:3907: in put
    await self._begin_copy(local_fs, self, localpaths, remotepath, 'put',
/build/asyncssh-2.13.1/asyncssh/sftp.py:3670: in _begin_copy
    srcattrs = await srcfs.stat(srcpath)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    async def stat(self, path: bytes) -> 'SFTPAttrs':
        """Get attributes of a local file or directory, following symlinks"""

>       return SFTPAttrs.from_local(os.stat(_to_local_path(path)))
E       OSError: [Errno 84] Invalid or incomplete multibyte or wide character: b'\xff'

/build/asyncssh-2.13.1/asyncssh/sftp.py:7347: OSError
------------------------------ Captured log call -------------------------------
INFO     asyncssh:logging.py:92 Opening SSH connection to 127.0.0.1, port 41343
INFO     asyncssh:logging.py:92 [conn=1692] Accepted SSH client connection
INFO     asyncssh:logging.py:92 [conn=1692]   Local address: 127.0.0.1, port 41343
INFO     asyncssh:logging.py:92 [conn=1692]   Peer address: 127.0.0.1, port 43100
INFO     asyncssh:logging.py:92 [conn=1693] Connected to SSH server at 127.0.0.1, port 41343
INFO     asyncssh:logging.py:92 [conn=1693]   Local address: 127.0.0.1, port 43100
INFO     asyncssh:logging.py:92 [conn=1693]   Peer address: 127.0.0.1, port 41343
INFO     asyncssh:logging.py:92 [conn=1693] Beginning auth for user guest
INFO     asyncssh:logging.py:92 [conn=1692] Beginning auth for user guest
INFO     asyncssh:logging.py:92 [conn=1692] Auth for user guest succeeded
INFO     asyncssh:logging.py:92 [conn=1693] Auth for user guest succeeded
INFO     asyncssh:logging.py:92 [conn=1693, chan=0] Requesting new SSH session
INFO     asyncssh:logging.py:92 [conn=1692, chan=0] New SSH session requested
INFO     asyncssh:logging.py:92 [conn=1693, chan=0]   Subsystem: sftp
INFO     asyncssh:logging.py:92 [conn=1692, chan=0]   Subsystem: sftp
INFO     asyncssh:logging.py:92 [conn=1692, chan=0] Closing channel
INFO     asyncssh:logging.py:92 [conn=1693, chan=0] Received channel close
INFO     asyncssh:logging.py:92 [conn=1693, chan=0] Channel closed
INFO     asyncssh:logging.py:92 [conn=1692, chan=0] Received channel close
INFO     asyncssh:logging.py:92 [conn=1693, chan=0] Closing channel
INFO     asyncssh:logging.py:92 [conn=1693] Closing connection
INFO     asyncssh:logging.py:92 [conn=1693] Sending disconnect: Disconnected by application (11)
INFO     asyncssh:logging.py:92 [conn=1692, chan=0] Channel closed
INFO     asyncssh:logging.py:92 [conn=1693] Connection closed
____________________ _TestSFTPAsync.test_copy_invalid_name _____________________

self = <tests.test_sftp._TestSFTPAsync testMethod=test_copy_invalid_name>
sftp = <asyncssh.sftp.SFTPClient object at 0x7ffff2dd8eb0>

    @sftp_test
    async def test_copy_invalid_name(self, sftp):
        """Test copying a file with an invalid name over SFTP"""

        for method in ('get', 'put', 'copy', 'mget', 'mput', 'mcopy'):
            with self.subTest(method=method):
                with self.assertRaises((FileNotFoundError, SFTPNoSuchFile,
                                        SFTPFailure, UnicodeDecodeError)):
>                   await getattr(sftp, method)(b'\xff')

/build/asyncssh-2.13.1/tests/test_sftp.py:873:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/build/asyncssh-2.13.1/asyncssh/sftp.py:3907: in put
    await self._begin_copy(local_fs, self, localpaths, remotepath, 'put',
/build/asyncssh-2.13.1/asyncssh/sftp.py:3670: in _begin_copy
    srcattrs = await srcfs.stat(srcpath)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    async def stat(self, path: bytes) -> 'SFTPAttrs':
        """Get attributes of a local file or directory, following symlinks"""

>       return SFTPAttrs.from_local(os.stat(_to_local_path(path)))
E       OSError: [Errno 84] Invalid or incomplete multibyte or wide character: b'\xff'

/build/asyncssh-2.13.1/asyncssh/sftp.py:7347: OSError
------------------------------ Captured log call -------------------------------
INFO     asyncssh:logging.py:92 Opening SSH connection to 127.0.0.1, port 34555
INFO     asyncssh:logging.py:92 [conn=2104] Accepted SSH client connection
INFO     asyncssh:logging.py:92 [conn=2104]   Local address: 127.0.0.1, port 34555
INFO     asyncssh:logging.py:92 [conn=2104]   Peer address: 127.0.0.1, port 47440
INFO     asyncssh:logging.py:92 [conn=2105] Connected to SSH server at 127.0.0.1, port 34555
INFO     asyncssh:logging.py:92 [conn=2105]   Local address: 127.0.0.1, port 47440
INFO     asyncssh:logging.py:92 [conn=2105]   Peer address: 127.0.0.1, port 34555
INFO     asyncssh:logging.py:92 [conn=2105] Beginning auth for user guest
INFO     asyncssh:logging.py:92 [conn=2104] Beginning auth for user guest
INFO     asyncssh:logging.py:92 [conn=2104] Auth for user guest succeeded
INFO     asyncssh:logging.py:92 [conn=2105] Auth for user guest succeeded
INFO     asyncssh:logging.py:92 [conn=2105, chan=0] Requesting new SSH session
INFO     asyncssh:logging.py:92 [conn=2104, chan=0] New SSH session requested
INFO     asyncssh:logging.py:92 [conn=2105, chan=0]   Subsystem: sftp
INFO     asyncssh:logging.py:92 [conn=2104, chan=0]   Subsystem: sftp
INFO     asyncssh:logging.py:92 [conn=2104, chan=0] Closing channel
INFO     asyncssh:logging.py:92 [conn=2105, chan=0] Received channel close
INFO     asyncssh:logging.py:92 [conn=2105, chan=0] Channel closed
INFO     asyncssh:logging.py:92 [conn=2104, chan=0] Received channel close
INFO     asyncssh:logging.py:92 [conn=2105, chan=0] Closing channel
INFO     asyncssh:logging.py:92 [conn=2105] Closing connection
INFO     asyncssh:logging.py:92 [conn=2105] Sending disconnect: Disconnected by application (11)
INFO     asyncssh:logging.py:92 [conn=2104, chan=0] Channel closed
INFO     asyncssh:logging.py:92 [conn=2105] Connection closed
=========================== short test summary info ============================
FAILED tests/test_sftp.py::_TestSFTP::test_copy_invalid_name - OSError: [Errno 84] Invalid or incomplete multibyte or wide character: b'\xff'
FAILED tests/test_sftp.py::_TestSFTPAsync::test_copy_invalid_name - OSError: [Errno 84] Invalid or incomplete multibyte or wide character: b'\xff'
==== 2 failed, 1457 passed, 11 skipped, 32 deselected in 626.26s (0:10:26) =====
```

**Comment by ronf** [comment]
Thanks for the report! Unfortunately, I'm not able to reproduce the problem here. The unit tests all run cleanly for me, on multiple versions of Linux, macOS, and Windows. What environment are you running on when you are seeing these failures?

**Comment by mweinelt** [comment]
I can reproduce these exceptions when building on a ZFS volume. Are you possibly building on one? @dotlambda

**Comment by mweinelt** [comment]
Yeah, filename normalization on ZFS is a curse and should not be enabled on machines that build code from source.

It would be appreciated if `OSError` would be caught as well, since the exception does indicate it is unhappy about the filename, though not in a way you tested, but hopefully in a way that makes sense for the test to expect.

**Comment by dotlambda** [comment]
> Are you possibly building on one?

I am indeed. However it's not possible to change the normalization property after the dataset was created :/

**Comment by ronf** [comment]
It would be appreciated if OSError would be caught as well, since the exception does indicate it is unhappy about the filename, though not in a way you tested, but hopefully in a way that makes sense for the test to expect.
That seems reasonable. It's a shame that they chose to map this to OSError instead of UnicodeDecodeError, though. Catching OSError means potentially catching a bunch of other failures that wouldn't be expected here::

      ├── OSError
      │    ├── BlockingIOError
      │    ├── ChildProcessError
      │    ├── ConnectionError
      │    │    ├── BrokenPipeError
      │    │    ├── ConnectionAbortedError
      │    │    ├── ConnectionRefusedError
      │    │    └── ConnectionResetError
      │    ├── FileExistsError
      │    ├── FileNotFoundError
      │    ├── InterruptedError
      │    ├── IsADirectoryError
      │    ├── NotADirectoryError
      │    ├── PermissionError
      │    ├── ProcessLookupError
      │    └── TimeoutError

However, since this only affects this one test case and not actual package behavior, I think that's ok.

This is now available as commit 8c0b42c in the "develop" branch.

**Comment by ronf** [comment]
This fix is now available in AsyncSSH 2.13.2.
